### ISSUE TITLE
Eliminate dependency on ImageScience library

### ISSUE THREAD

**Issue by ctrueden** [fiji/Trainable_Segmentation#4]
Unfortunately, the [ImageScience library's license](http://fiji.sc/ImageScience_License) does not allow it to be consumed as a dependency from GPL code.

**Comment by iarganda** [comment]
@ctrueden can I close this issue? Is this solution satisfactory for you?

**Comment by ctrueden** [comment]
__TL;DR: I will submit a PR later today that addresses this issue to my satisfaction.__

@iarganda Legally, ImageScience has its ducks in a row now, as a result of the above discussion. But the situation with TWS is trickier, and not solved as things stand.

In case you haven't read them yet, these FAQ entries discuss the relevant issue:
* [Can I write free software that uses non-free libraries?](https://www.gnu.org/licenses/gpl-faq.html#FSWithNFLibs)
* [What legal issues come up if I use GPL-incompatible libraries with GPL software?](https://www.gnu.org/licenses/gpl-faq.html#GPLIncompatibleLibs)

The FAQ is a little self-referential and obtuse, but the gist is that from a legal standpoint, the TWS project would need to have a modified GPLv3 license with an extra clause:

> Additional permission under GNU GPL version 3 section 7
>
> If you modify this Program, or any covered work, by linking or combining it with ImageScience (http://fiji.sc/ImageScience), containing parts covered by the terms of the ImageScience license (http://fiji.sc/ImageScience_License), the licensors of this Program grant you additional permission to convey the resulting work.

Unfortunately, such a blurb would be highly misleading, because it implies that it would be legal to modify TWS, then redistribute the modified version of TWS + dependencies outside Fiji, when in fact such redistribution would be illegal under the terms of the ImageScience license (and its limited exception allowing redistribution within Fiji only).

I am sorry to make trouble here. But as things stand, this issue is not resolved from a legal standpoint, and it makes alarm bells go off in my head. We could choose to ignore the issue, and in practice no one would care with respect to the current distribution mechanisms. But it would be a legal problem if:

* Someone wanted to distribute Fiji as part of the Debian Linux distribution. TWS (and ImageScience and the BIG plugins) would all need to be excluded from that distribution mechanism.
* Someone wanted to add TWS as a dependency of their program which is distributed outside of Fiji.

And I am fairly certain there are other scenarios which could crop up in the future, but haven't explicitly occurred to me. The upshot is that it makes it much more difficult for people to legally modify and redistribute the TWS plugin, which goes against our [open source science principles](http://imagej.net/Open_Source).

From a technical standpoint, TWS uses ImageScience only for computation of certain features. 
Hence, we can make the ImageScience dependency optional, by making the ImageScience-driven feature computations optional. I am working on a patch now, and will file a PR later today.

In the longer term, it would be awesome to add ops computing these features to the ImageJ OPS library, and fully remove the ImageScience dependency. But that is a project for another day.

**Comment by iarganda** [comment]
OK, thanks for the clear explanation! I agree that using ImageJ OPS features substituting the ImageScience ones is the right way to go. Hopefully soon!
